gcovr/gcov.py

```diff
@@ -90,170 +85,274 @@ def is_non_code(code):
+        '               {segments}\n'
+        '  source_fname {source_fname}\n'
+        '  root         {root_dir}\n'
+        # '  common_dir   {common_dir}\n'
+        # '  subdir       {subdir}\n'
+        '  fname        {fname}',
+        currdir=currdir, data_fname=data_fname, segments=segments,
+        source_fname=source_fname, root_dir=root_dir,
+        # common_dir=common_dir, subdir=subdir,
+        fname=fname)
+
+    return fname
+
+
+def guess_source_file_name_via_aliases(gcovname, currdir, data_fname):
+    common_dir = os.path.commonprefix([data_fname, currdir])
+    fname = aliases.unalias_path(os.path.join(common_dir, gcovname))
+    if os.path.exists(fname):
+        return fname
+
+    initial_fname = fname
+
+    data_fname_dir = os.path.dirname(data_fname)
+    fname = aliases.unalias_path(os.path.join(data_fname_dir, gcovname))
+    if os.path.exists(fname):
+        return fname
+
+    # @latk-2018: The original code is *very* insistent
+    # on returning the inital guess. Why?
+    return initial_fname
+
+
+def guess_source_file_name_heuristics(
+        gcovname, currdir, root_dir, starting_dir, source_fname):
+
+    # gcov writes filenames with '/' path seperators even if the OS
+    # separator is different, so we replace it with the correct separator
+    gcovname = gcovname.replace('/', os.sep)
+
+    # 0. Try using the current working directory as the source directory
+    fname = os.path.join(currdir, gcovname)
+    if os.path.exists(fname):
+        return fname
+
+    # 1. Try using the path to common prefix with the root_dir as the source directory
+    fname = os.path.join(root_dir, gcovname)
+    if os.path.exists(fname):
+        return fname
+
+    # 2. Try using the starting directory as the source directory
+    fname = os.path.join(starting_dir, gcovname)
+    if os.path.exists(fname):
+        return fname
+
+    # 3. Try using the path to the gcda file as the source directory
+    source_fname_dir = os.path.dirname(source_fname)
+    fname = os.path.join(source_fname_dir, os.path.basename(gcovname))
+    return fname
+
+
+class GcovParser(object):
+    def __init__(self, fname, logger):
+        self.logger = logger
+        self.excluding = []
+        self.noncode = set()
+        self.uncovered = set()
+        self.uncovered_exceptional = set()
+        self.covered = dict()
+        self.branches = dict()
+        # self.first_record = True
+        self.fname = fname
+        self.lineno = 0
+        self.last_code_line = ""
+        self.last_code_lineno = 0
+        self.last_code_line_excluded = False
+        self.unrecognized_lines = []
+        self.deferred_exceptions = []
+        self.last_was_specialization_section_marker = False
+
+    def parse_all_lines(self, lines, exclude_unreachable_branches):
+        for line in lines:
+            try:
+                self.parse_line(line, exclude_unreachable_branches)
+            except Exception as ex:
```

Comment:
This shall probably be an opt-in behavior. I'd rather have `gcovr` fail with an error message and reporting we can opt-in to ignore those by adding flag `--ignore-parsing-errors` or something like that.

Reply:
You're absolutely right. I'll add a commit to this PR. Thank you for spotting this issue!

Reply:
Ok, I've added a --gcov-ignore-parse-errors flag that is default-off.

If the flag is enabled, gcovr will

 - print all offending lines,
 - print all caught exceptions,
 - and continue processing.

If the flag is not enabled (the default), gcovr does not continue but

 - also prints offending lines and caught exceptions,
 - then displays an error message that the flag can be enabled,
 - then rethrows the first exception if one was caught in order to display its traceback
 - or `sys.exit(1)` if no exceptions were caught.

In all cases, the errors are processed only *after* a the complete file has been parsed.